src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -109,14 +117,124 @@ protected Object readResolve() throws ObjectStreamException {
+            Map<FlowExecutionOwner, Map<String, Boolean>> startNodesAreClosedByFlow = FlowL.startNodesAreClosedByFlow();
+            LOGGER.log(Level.FINER, "for {0}, startNodesAreClosedByFlow={1}", new Object[] {this, startNodesAreClosedByFlow});
+            Map<String, Boolean> startNodesAreClosed = startNodesAreClosedByFlow.get(exec.getOwner());
+            if (startNodesAreClosed != null) {
+                Boolean closed = startNodesAreClosed.get(id);
+                if (closed != null) { // quick version
+                    LOGGER.log(Level.FINER, "quick closed={0}", closed);
+                    return !closed;
+                } else {
+                    LOGGER.log(Level.FINER, "no record of {0} in {1}, presumably GraphListener not working", new Object[] {this, exec});
+                }
+            } else {
+                LOGGER.log(Level.FINER, "no record of {0}, either FlowExecutionListener not working or it is already complete", exec);
+            }
+        }
+        List<FlowNode> currentHeads = exec.getCurrentHeads();
+        if (currentHeads.contains(this)) {
+            LOGGER.log(Level.FINER, "{0} is a current head", this);
+            return true;
+        }
+        if (currentHeads.size() == 1 && currentHeads.get(0) instanceof FlowEndNode) { // i.e., exec.isComplete()
+            LOGGER.log(Level.FINER, "{0} is complete", exec);
+            return false;
+        }
+        if (this instanceof BlockStartNode) {
+            // Fallback (old workflow-job, old workflow-cps):
+            LOGGER.log(Level.FINER, "slow currentHeads={0}", currentHeads);
+            for (FlowNode headNode : currentHeads) {
+                if (new LinearBlockHoppingScanner().findFirstMatch(headNode, Predicates.equalTo(this)) != null) {
+                    LOGGER.finer("slow match");
+                    return true;
+                }
+            }
+            LOGGER.finer("slow no match");
+            return false;
+        }
+        LOGGER.log(Level.FINER, "{0} is not a current head nor a start node", this);
+        return false;
+    }
+    /**
+     * Cache of known block start node statuses.
+     * Keys are running executions ~ builds.
+     * Values are maps from {@link BlockStartNode#getId} to whether the corresponding {@link BlockEndNode} has been encountered.
+     * For old {@code workflow-job}, the top-level entries will be missing;
+     * for old {@code workflow-cps}, the second-level entries will be missing.
+     */
+    @Restricted(DoNotUse.class)
+    @Extension public static final class GraphL implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            LOGGER.finer("GraphListener working");
+            if (node instanceof BlockStartNode || node instanceof BlockEndNode) {
+                Map<String, Boolean> startNodesAreClosed = FlowL.startNodesAreClosedByFlow().get(node.getExecution().getOwner());
+                if (startNodesAreClosed != null) {
```

Comment:
Mention this is safe to do because the FlowExecutionListener has ensured that the FlowExecution is only null if flow is done (otherwise always set for all running builds).